File: JavaScript/6js 1/bouncingBall.js
/*
|--------------------------------------------------------------------------
| Bouncing Balls
|--------------------------------------------------------------------------
|
*/

function bouncingBall(h, bounce, window) {
   if (h <= 0 || bounce <= 0 || bounce >= 1 || window >= h) return -1;
   let res = 1;
   while (h > window) {
      h *= bounce;
      if (h > window) res += 2;
   }
   return res;
}


const bouncingBall = (h, b, w) =>
   b >= 1 || h < w || b <= 0 ? -1 : Math.ceil(Math.log(w / h) / Math.log(b)) * 2 - 1;


function bouncingBall(h, bounce, window) {
   if (h <= 0 || bounce <= 0 || bounce >= 1 || window >= h) {
      return -1;
   }
   let newHeight = h * bounce;
   return bouncingBall(newHeight, bounce, window) + 2;
}


console.log(bouncingBall(3.0, 0.66, 1.5), 3);
console.log(bouncingBall(30.0, 0.66, 1.5), 15);
console.log(bouncingBall(3.0, 1.0, 1.5), -1);

File: Python/6py 1/bouncingBall.py
"""
|--------------------------------------------------------------------------
| Bouncing Balls
|--------------------------------------------------------------------------
|
"""

def bouncingBall(h, bounce, window):
   if h <= 0 or bounce <= 0 or bounce >= 1 or window >= h: return -1
   res = 1
   while h > window:
      h *= bounce
      if h > window: res += 2
   return res


def bouncingBall(h, bounce, window):
   if h <= 0 or bounce <= 0 or bounce >= 1 or window >= h: return -1
   return 2 + bouncingBall(h * bounce, bounce, window)


from math import log
def bouncingBall(h, bounce, window):
   return 2*int(log(window/h)/log(bounce)) + 1 if h > 0 and 0 < bounce < 1 and window < h else -1


print(bouncingBall(3.0, 0.66, 1.5), 3)
print(bouncingBall(30.0, 0.66, 1.5), 15)
print(bouncingBall(3.0, 1.0, 1.5), -1)